content_service.go

```diff
@@ -302,28 +298,34 @@ func (con *contentService) getHistory(ctx context.Context,
 
 func (con *contentService) getDiff(ctx context.Context,
 	projectName, repoName, from, to string, query *Query) (*Change, int, error) {
+
+	// validate query
 	if query == nil {
 		return nil, UnknownHttpStatusCode, errors.New("query should not be nil")
 	}
-
-	path := query.Path
-	if len(path) == 0 {
+	if len(query.Path) == 0 {
 		return nil, UnknownHttpStatusCode, errors.New("the path should not be empty")
 	}
-	if !strings.HasPrefix(path, "/") {
-		path = "/" + path
+
+	// build relative url
+	u, err := url.Parse(path.Join(
+		defaultPathPrefix,
+		urlPartProjects, projectName,
+		urlPartRepos, repoName,
+		urlPartActionCompare,
+	))
+	if err != nil {
+		return nil, UnknownHttpStatusCode, err
 	}
 
-	u := fmt.Sprintf("%vprojects/%v/repos/%v/compare", defaultPathPrefix, projectName, repoName)
-	v := &url.Values{}
-	v.Set("path", path)
-	if query != nil && query.Type == JSONPath {
-		if err := setJSONPaths(v, path, query.Expressions); err != nil {
-			return nil, UnknownHttpStatusCode, err
-		}
+	// build query params
+	q := u.Query()
+	setPath(&q, path.Join("/", query.Path))
+	if err := setJSONPaths(&q, query); err != nil {
```

Comment:
Shouldn't we do this only when `query != nil`?

Reply:
`query` is always not nil. Above code check for that. and `q` is always not nil too.

setJSONPaths may return non nil error when user `query` type is json but path not end with `json`

Reply:
Therefore, I guess code logic here is correct

Reply:
Ah yes, right. 😉